sdk/keyvault/keyvault-keys/test/tracing.test.ts

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { KeyClient, JsonWebKey } from "../src";
+import * as assert from "assert";
+import { TestTracer, setTracer, OperationTracingOptions } from "@azure/core-tracing";
+import { Context as MochaTestContext, Test } from "mocha";
+import { env, Recorder, isRecordMode, isPlaybackMode, delay } from "@azure/test-utils-recorder";
+import { authenticate } from "./utils/testAuthentication";
+import TestClient from "./utils/testClient";
+import { testPollerProperties } from "./utils/recorderUtils";
+import { isNode } from "@azure/core-http";
+
+// This test file confirms that all of our methods that create tracing spans
+// are effectively doing so.
+
+/**
+ * A catch-all type for the SpanGraph root and its nodes.
+ */
+interface TestRecursiveGraph {
+  name?: string;
+  roots?: TestRecursiveGraph[];
+  children?: TestRecursiveGraph[];
+}
+
+/**
+ * Gets the names of all of the nodes of a SpanGraph from @azure/core-tracing
+ */
+function getNames(graph: TestRecursiveGraph): string[] {
+  const { name, roots, children = roots } = graph;
+  if (!children || !children!.length) {
+    return [];
+  }
+  const childrenNames = children.map(getNames).flat();
+  return [name!, ...childrenNames].filter((x) => x);
+}
+
+export type TracingTestFunc = (
+  this: MochaTestContext,
+  tracingOptions: OperationTracingOptions
+) => Promise<void>;
+
+interface TestTracingNodeNames {
+  (nodeNames: string[], test: TracingTestFunc): Test;
+  skip: (nodeNames: string[], test: TracingTestFunc) => Test;
+  only: (nodeNames: string[], test: TracingTestFunc) => Test;
+}
+
+function generateTracingTest(nodeNames: string[], test: TracingTestFunc): Mocha.Func {
+  return async function() {
+    const tracer = new TestTracer();
+    setTracer(tracer);
+    const rootSpan = tracer.startSpan("root");
+
+    await test.call(this, {
+      spanOptions: { parent: rootSpan.context() }
+    });
+
+    const graph = tracer.getSpanGraph(rootSpan.context().traceId);
+    const names = getNames(graph);
+    const uniqueNames = names.reduce(
+      (unique, next) => (unique.includes(next) ? unique : [...unique, next]),
+      ["root"]
+    );
+    assert.deepEqual(["root", ...nodeNames], uniqueNames);
```

Comment:
Technically, the hierarchy of spans is important, not just the order/list of spans. Improperly parented spans makes the telemetry data less useful. This is why TestTracer returns a graph rather than an array.

Reply:
The issue here is that LRO graphs can be very, very large. In LRO cases, we don't need to know how many repetitions we have of the requests.

Also, although soundness of the structure makes sense to debug what's going on, it seems sufficient to me to check that these records exist in the structure, then trust that the rest of identity is already tested and working.

Reply:
It sounds like LROs need some kind of custom validator, where you can express "this child span is expected 1 or more times"

Reply:
@xirzec what do you think if I make new issues for core-tracing to define validators that could simplify tests?

Reply:
That is a good idea